sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala

```diff
@@ -0,0 +1,411 @@
+      "Cannot find data for output column", "'x'"))
+  }
+
+  test("Append.byName: fail canWrite check") {
+    val parsedPlan = AppendData.byName(table, widerTable)
+
+    assertNotResolved(parsedPlan)
+    assertAnalysisError(parsedPlan, Seq(
+      "Cannot write", "'table-name'",
+      "Cannot safely cast", "'x'", "'y'", "DoubleType to FloatType"))
+  }
+
+  test("Append.byName: insert safe cast") {
+    val x = table.output.toIndexedSeq(0)
+    val y = table.output.toIndexedSeq(1)
+
+    val parsedPlan = AppendData.byName(widerTable, table)
+    val expectedPlan = AppendData.byName(widerTable,
+      Project(Seq(
+        Alias(Cast(x, DoubleType, Some(conf.sessionLocalTimeZone)), "x")(),
+        Alias(Cast(y, DoubleType, Some(conf.sessionLocalTimeZone)), "y")()),
+        table))
+
+    assertNotResolved(parsedPlan)
+    checkAnalysis(parsedPlan, expectedPlan)
+    assertResolved(expectedPlan)
+  }
+
+  test("Append.byName: fail extra data fields") {
+    val query = TestRelation(StructType(Seq(
+      StructField("x", FloatType),
+      StructField("y", FloatType),
+      StructField("z", FloatType))).toAttributes)
+
+    val parsedPlan = AppendData.byName(table, query)
+
+    assertNotResolved(parsedPlan)
+    assertAnalysisError(parsedPlan, Seq(
+      "Cannot write", "'table-name'", "too many data columns",
+      "Table columns: 'x', 'y'",
+      "Data columns: 'x', 'y', 'z'"))
+  }
+
+  test("Append.byName: multiple field errors are reported") {
+    val xRequiredTable = TestRelation(StructType(Seq(
+      StructField("x", FloatType, nullable = false),
+      StructField("y", DoubleType))).toAttributes)
+
+    val query = TestRelation(StructType(Seq(
+      StructField("x", DoubleType),
+      StructField("b", FloatType))).toAttributes)
+
+    val parsedPlan = AppendData.byName(xRequiredTable, query)
+
+    assertNotResolved(parsedPlan)
+    assertAnalysisError(parsedPlan, Seq(
+      "Cannot write incompatible data to table", "'table-name'",
+      "Cannot safely cast", "'x'", "DoubleType to FloatType",
+      "Cannot write nullable values to non-null column", "'x'",
+      "Cannot find data for output column", "'y'"))
+  }
+
+  test("Append.byPosition: basic behavior") {
+    val query = TestRelation(StructType(Seq(
+      StructField("a", FloatType),
+      StructField("b", FloatType))).toAttributes)
+
+    val a = query.output.toIndexedSeq(0)
+    val b = query.output.toIndexedSeq(1)
+
+    val parsedPlan = AppendData.byPosition(table, query)
+    val expectedPlan = AppendData.byPosition(table,
+      Project(Seq(
+        Alias(Cast(a, FloatType, Some(conf.sessionLocalTimeZone)), "x")(),
+        Alias(Cast(b, FloatType, Some(conf.sessionLocalTimeZone)), "y")()),
+        query))
+
+    assertNotResolved(parsedPlan)
+    checkAnalysis(parsedPlan, expectedPlan, caseSensitive = false)
+    assertResolved(expectedPlan)
+  }
+
+  test("Append.byPosition: case does not fail column resolution") {
```

Comment:
do we need this test? In "Append.byPosition: basic behavior" we proved that we can do append even the column names are different.